### PR TITLE
Translation: Implement override of get_message

### LINKED ISSUE #1926: Allow to extend Translation class in the script or GDNative
### Describe the project you are working on

FreeOrion Godot client, alternative client with Godot UI https://github.com/freeorion/freeorion/pull/3235

### Describe the problem or limitation you are having in your project

We use own internationalization system and want to use already existing system in the Godot project.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Allow `TranslationServer` to call `get_message` of `Translation` object via `Object::call` instead of virtual function call.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Call for the game's i18n code: https://github.com/o01eg/freeorion/blob/43ca1d3d17e8baa946cd03e872293ecd5560c7a8/godot/GodotI18n.cpp#L42

Initialization of internationalization: https://github.com/o01eg/freeorion/blob/43ca1d3d17e8baa946cd03e872293ecd5560c7a8/godot/gdfreeorion.cpp#L196-L201

### If this enhancement will not be used often, can it be worked around with a few lines of script?

The only idea is to generate CSV from the game's translation files which will have to be re-imported each time main translation change.

### Is there a reason why this should be core and not an add-on in the asset library?

Asset library don't have access to translation internals.

### PR DESCRIPTION
**Pull request by o01eg**
This PR switches `Translation::get_message` and `Translation::get_plural_message` virtual call to GDScript compatible call. Also it introduces unit test for translations.

Closes https://github.com/godotengine/godot-proposals/issues/1926

### CODE REVIEW

File: tests/test_translations.h
Comment (Calinou) [inline tests/test_translations.h:56]: The file path should be relative to the binary, so that tests pass regardless of the folder you run the Godot binary from.

See https://github.com/godotengine/godot/blob/master/tests/test_pck_packer.h#L82-L86 for an example of doing this.

**Review by Calinou — DISMISSED** [review]

File: tests/test_translations.h
Comment (o01eg) [reply tests/test_translations.h:56]: Ok, I just copied path from `tests/test_file_access.h`. I'll change it.

Comment (Xrayez) [reply tests/test_translations.h]: My bad, should be fixed there as well eventually.

**Comment by o01eg** [comment]
Are there any news regarding this PR?

File: tests/test_translations.h
Comment (akien-mga) [inline tests/test_translations.h:57]: Use `TestUtils::get_data_path()` for that.

Comment (o01eg) [reply tests/test_translations.h]: Fixed.

File: tests/test_translations.h
Comment (Xrayez) [inline tests/test_translations.h:45]: I'm a bit concerned about instantiation of another `TranslationServer` because it's likely already setup in `test_main()`. Instantiating `TranslationServer` may lead to an override of the main singleton.

You can try to remove this, include translation server and use `TranslationServer::get_singleton()` instead, and it should work I think.

File: tests/test_translations.h
Comment (Xrayez) [inline tests/test_translations.h:64]: That's what I'm talking about, you either have to use `TranslationServer::get_singleton()` or instantiate another server (which goes unused in this test case it seems), but I'd just use `TranslationServer::get_singleton()`.

File: core/string/translation.cpp
Comment (Xrayez) [inline core/string/translation.cpp:1133]: Is it possible to remove `const` from `const Set<Ref<Translation>>::Element *E` here? Then you probably wouldn't need to `const_cast`.

File: tests/test_translations.h
Comment (o01eg) [reply tests/test_translations.h]: I've tried to remove it but it looks like singleton wasn't initialized.

**Comment by Xrayez** [comment]
Ok, crashes now: https://github.com/godotengine/godot/pull/44670/checks?check_run_id=1952034051#step:9:59.

I suspect this is because `TranslationServer` is not actually instantiated in the test environment. So, we have two options:
- Instantiate and load translations in `test_main()`.
- Make a local instance of `TranslationServer`.

So, the original:
```
TranslationServer *translation_server = nullptr;
translation_server = memnew(TranslationServer);
translation_server->setup(); //register translations, load them, etc.
translation_server->load_translations();
```

may be acceptable way to do this, but it would only work fine for the test case itself, but may cause problems down the road if someone would want to test `TranslationServer` anywhere else.

File: core/string/translation.cpp
Comment (o01eg) [reply core/string/translation.cpp:1133]: No, `call` is non-const function so I cannot call it otherwise.

**Comment by o01eg** [comment]
@Xrayez Should I only initialize `TranslationServer` and it will be finalized itself. Or I should finalize it after `test_context.run()` too?

**Comment by Xrayez** [comment]
Sorry, I've been working on resolving this issue, see #46313.

File: tests/test_main.cpp
Comment (Xrayez) [inline tests/test_main.cpp:95]: Should be handled by #46313, will need to rebase later.

And sorry, I forgot that the test environment is setup in `Main::test_setup()`, not `test_main()`.

**Comment by o01eg** [comment]
I suppose it better to add files into `.gitignore` because it possible on Windows I cannot delete them while tests are executing.

File: tests/test_main.cpp
Comment (o01eg) [reply tests/test_main.cpp]: I rebased this PR over #46313

**Review by Xrayez — APPROVED** [review]
Tests-wise, should be ok now, but need rebase after #46313.

Regarding `Translation` changes, I think this shouldn't hurt anything.

**Comment by o01eg** [comment]
Rebased. Could it be also backported to 3.2 branch?

**Comment by Xrayez** [comment]
I think this can be done, but without tests because 3.2 doesn't have tests implemented as in 4.0 (`master`).

Also, by rebasing I also meant that commits must be squashed into one before merging, see [PR workflow in Godot](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase). 🙂

Akien is responsible for merging and cherry-picking PRs to 3.2.

**Comment by o01eg** [comment]
Squashed it.

**Review by Xrayez — APPROVED** [review]

**Comment by o01eg** [comment]
Does something else is needed to accept this PR?

**Comment by akien-mga** [comment]
It needs a review by core engine maintainers (not sure why the previous review request was discarded). I don't know if the changes done to `core/string/translation.cpp` are the proper way to support the use case in godotengine/godot-proposals#1926.

**Comment by o01eg** [comment]
Rebased and moved test to recently introduced translation tests.

File: tests/test_translation.h
Comment (akien-mga) [inline tests/test_translation.h:38]: This and the added test should be enclosed in `#ifdef TOOLS_ENABLED`. I think currently we only build tests with the editor so it's always true, but in the future we might want to make it possible to build tests for the export templates too.

Comment (o01eg) [reply tests/test_translation.h:39]: Added definition check.

File: core/string/translation.cpp
Comment (vnen) [inline core/string/translation.cpp:1164]: In general we do not allow scripts to override methods from the engine classes. The common pattern for these cases is to register a virtual method for scripting that can be overridden, and on the engine side the implementation (in this case `get_message()` and `get_plural_message()`) should call the virtual method if there's a script attached.

Comment (vnen) [reply core/string/translation.cpp:1164]: As an example, BaseButton has a `_pressed()` virtual method that will be called when the button is pressed:

https://github.com/godotengine/godot/blob/382ddd497a732812356fccbe9642d86848a4da9d/scene/gui/base_button.cpp#L123-L129

Note that the fact that the C++ function having the same name as the virtual function is just coincidence, since the C++ function isn't bound to scripting only the virtual version is:

https://github.com/godotengine/godot/blob/382ddd497a732812356fccbe9642d86848a4da9d/scene/gui/base_button.cpp#L426

Comment (akien-mga) [reply core/string/translation.cpp:1164]: You can find more examples of this kind of virtual scripting methods in #48746 as many have been renamed to harmonize their names.

So here you should likely bind `_get_message()` and `_get_plural_message()` as virtual methods additionally to the C++ functions `get_message()` and `get_plural_message()` (which are also public and bound IINM). Then the implementation of `get_message()` would call `_get_message()` if defined in the script:

```gdscript
StringName Translation::get_message(const StringName &p_src_text, const StringName &p_context) const {
	if (get_script_instance()) {
		return get_script_instance()->call("_get_message", p_src_text, p_context);
	}

	if (p_context != StringName()) {
		WARN_PRINT("Translation class doesn't handle context. Using context in get_message() on a Translation instance is probably a mistake. \nUse a derived Translation class that handles context, such>
	}

	const Map<StringName, StringName>::Element *E = translation_map.find(p_src_text);
	if (!E) {
		return StringName();
	}

	return E->get();
}
```

Comment (vnen) [reply core/string/translation.cpp:1164]: In this case you should also check if the `_get_message()` function is defined in the script.

Comment (o01eg) [reply core/string/translation.cpp]: Rebased and changed code to call script instance methods.

Comment (akien-mga) [reply core/string/translation.cpp]: Looks good, but it still needs actually binding these as virtual methods as documented here: https://github.com/godotengine/godot/pull/44670#discussion_r662508489
Otherwise users won't know they exist, and then they serve little purpose. This will also add them to the documentation so you can fill their descriptions.

Comment (o01eg) [reply core/string/translation.cpp]: As I understand BIND_VMETHOD supposed to be removed in https://github.com/godotengine/godot/pull/51983

Comment (akien-mga) [reply core/string/translation.cpp]: Ah you're right, but just removing it is not sufficient, it needs to be replaced by the new syntax. Search e.g. `_get_connection_line` in #51983 for an example of what's needed.

I suggest splitting this in `master` and `3.x` PRs as we'll probably want the ability to override `get_message` in `3.x`, but there it should use `BIND_VMETHOD` and `call` like you do here.

Comment (o01eg) [reply core/string/translation.cpp]: Should I open PR with `BIND_VMETHOD` on `3.x` branch?

Comment (akien-mga) [reply core/string/translation.cpp]: Yes, please do :)

Comment (o01eg) [reply core/string/translation.cpp]: Changed this PR to `GDVIRTUAL_BIND`.

File: tests/test_translation.h
Comment (akien-mga) [inline tests/test_translation.h:181]: Missing newline at end of file.

File: core/string/translation.cpp
Comment (akien-mga) [inline core/string/translation.cpp:937]: Should match the non virtual arguments so that documentation is easier to parse for both methods.
```suggestion
	GDVIRTUAL_BIND(_get_plural_message, "src_message", "src_plural_message", "n", "context");
	GDVIRTUAL_BIND(_get_message, "src_message", "context");
```

File: tests/test_translation.h
Comment (o01eg) [reply tests/test_translation.h]: Fixed.

File: core/string/translation.h
Comment (o01eg) [reply core/string/translation.h]: Fixed.

File: core/string/translation.cpp
Comment (o01eg) [reply core/string/translation.cpp]: Fixed.

File: doc/classes/Translation.xml
Comment (akien-mga) [inline doc/classes/Translation.xml:19]:
```suggestion
				Virtual method to override [method get_message].
```

File: doc/classes/Translation.xml
Comment (akien-mga) [inline doc/classes/Translation.xml:29]:
```suggestion
				Virtual method to override [method get_plural_message].
```

File: tests/test_translation.h
Comment (akien-mga) [inline tests/test_translation.h:182]: There's no reason to change this file in this PR anymore.

File: core/string/translation.h
Comment (akien-mga) [inline core/string/translation.h:36]: Why is this include needed? The macros used in this PR seem to come from `core/object/object.h` and `core/object/gdvirtual.gen.inc`.

Comment (o01eg) [reply core/string/translation.h:36]: I searched for incomplete class `ScriptInstance` and found that include.

File: doc/classes/Translation.xml
Comment (o01eg) [reply doc/classes/Translation.xml]: Fixed.

File: doc/classes/Translation.xml
Comment (o01eg) [reply doc/classes/Translation.xml]: Fixed.

File: tests/test_translation.h
Comment (o01eg) [reply tests/test_translation.h]: Removed from PR.

File: core/string/translation.h
Comment (o01eg) [reply core/string/translation.h]: Changed to `core/object/object.h`

Comment (akien-mga) [reply core/string/translation.h]: Ah I see, the include for `ScriptInstance` should likely be added to `core/object/gdvirtual.gen.inc` itself.

`core/object/object.h` is probably not needed here as it's already included by `core/io/resource.h` (though it's not wrong to have it here either).

Comment (akien-mga) [reply core/string/translation.h]: > Ah I see, the include for ScriptInstance should likely be added to core/object/gdvirtual.gen.inc itself.

That's outside the scope of this PR though, I can have a look at this myself. Feel free to put back `script_language.h` here since it's needed.

**Review by akien-mga — APPROVED** [review]
Looks great! Thanks for doing all the necessary edits.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
